### PR TITLE
 Add a11y label to modal barrier

### LINKED ISSUE #13451: Dismissible ModalBarrier needs an a11y label
On iOS VoiceOver currently reads out garbage when a11y focusing on the ModalBarrier because it doesn't have a proper label.

Android doesn't have the problem because on this platform the barrier is not a11y focusable (the back button is used to dismiss).

### PR DESCRIPTION
**Pull request by goderbauer**
Breaking change announcement has been made here: https://groups.google.com/forum/#!topic/flutter-dev/FzUAHl2z4mo

Fixes https://github.com/flutter/flutter/issues/13451.

@HansMuller for the integration of translations.

### CODE REVIEW

File: packages/flutter_localizations/lib/src/l10n/material_ja.arb [hunk 1]
```diff
@@ -29,5 +29,6 @@
   "anteMeridiemAbbreviation": "AM",
   "postMeridiemAbbreviation": "PM",
   "timePickerHourModeAnnouncement": "時を選択",
-  "timePickerMinuteModeAnnouncement": "分を選択"
+  "timePickerMinuteModeAnnouncement": "分を選択",
+  "modalBarrierDismissLabel": "閉じる"
```
Comment (cbracken) [inline packages/flutter_localizations/lib/src/l10n/material_ja.arb]: LGTM for Japanese.

File: packages/flutter_localizations/lib/src/l10n/material_fr.arb [hunk 2]
```diff
@@ -30,5 +30,6 @@
   "anteMeridiemAbbreviation": "AM",
   "postMeridiemAbbreviation": "PM",
   "timePickerHourModeAnnouncement": "Sélectionnez les heures",
-  "timePickerMinuteModeAnnouncement": "Sélectionnez les minutes"
+  "timePickerMinuteModeAnnouncement": "Sélectionnez les minutes",
+  "modalBarrierDismissLabel": "Ignorer"
```
Comment (cbracken) [inline packages/flutter_localizations/lib/src/l10n/material_fr.arb]: LGTM for French.

File: packages/flutter/lib/src/widgets/modal_barrier.dart [hunk 3]
```diff
@@ -49,20 +50,30 @@ class ModalBarrier extends StatelessWidget {
   ///    [ModalBarrier] built by [ModalRoute] pages.
   final bool dismissible;
 
+  /// Semantics label used for the barrier if it is [dismissable].
+  ///
+  /// See also:
+  ///
+  ///  * [ModalRoute.barrierLabel], which controls this property for the
+  ///    [ModalBarrier] built by [ModalRoute] pages.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/modal_barrier.dart]: maybe see also something that tells you what a semantics label is? :-)

File: packages/flutter_localizations/lib/src/l10n/material_de_CH.arb [hunk 4]
```diff
@@ -18,7 +18,7 @@
   "selectedRowCountTitleOne": "1 Element ausgewählt",
   "selectedRowCountTitleOther": "$selectedRowCount Elemente ausgewählt",
   "cancelButtonLabel": "ABBRECHEN",
-  "closeButtonLabel": "SCHLIEẞEN",
+  "closeButtonLabel": "SCHLIESSEN",
```
Comment (HansMuller) [inline packages/flutter_localizations/lib/src/l10n/material_de_CH.arb]: I have been trying to get the internal translations team to update this value (with little success). Your change may be reverted when the next translations update happens.

File: packages/flutter/lib/src/widgets/modal_barrier.dart [hunk 5]
```diff
@@ -49,20 +50,30 @@ class ModalBarrier extends StatelessWidget {
...
+        excluding: !semanticsDismissable,
+        child: new GestureDetector(
+          onTapDown: (TapDownDetails details) {
+            if (dismissible)
+              Navigator.pop(context);
+          },
+          behavior: HitTestBehavior.opaque,
+          child: new Semantics(
+            label: semanticsDismissable ? semanticsLabel : null,
+            textDirection: semanticsDismissable ? Directionality.of(context) : null,
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/modal_barrier.dart]: start the build method with an assert that calls debugCheckHasDirectionality or whatever it's called

File: packages/flutter/lib/src/widgets/modal_barrier.dart [hunk 6]
```diff
@@ -118,11 +130,20 @@ class AnimatedModalBarrier extends AnimatedWidget {
   ///    [AnimatedModalBarrier] built by [ModalRoute] pages.
   final bool dismissible;
 
+  /// Semantics label used for the barrier if it is [dismissable].
+  ///
+  /// See also:
+  ///
+  ///  * [ModalRoute.barrierLabel], which controls this property for the
+  ///    [ModalBarrier] built by [ModalRoute] pages.
+  final String semanticsLabel;
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/modal_barrier.dart]: ditto earlier comment

File: packages/flutter/test/material/dialog_test.dart [hunk 7]
```diff
@@ -234,21 +234,28 @@ void main() {
...
-              },
-            );
-          },
+    await tester.pumpWidget(new Localizations(
+      locale: const Locale('en', 'US'),
+      delegates: <LocalizationsDelegate<dynamic>>[
+        DefaultWidgetsLocalizations.delegate,
+        DefaultMaterialLocalizations.delegate,
+      ],
+      child: new Directionality(
```
Comment (Hixie) [inline packages/flutter/test/material/dialog_test.dart]: you can remove the Directionality, Localizations adds one

File: packages/flutter/test/material/dropdown_test.dart [hunk 8]
```diff
@@ -63,18 +63,25 @@ class TestApp extends StatefulWidget {
...
-          },
+    return new Localizations(
+      locale: const Locale('en', 'US'),
+      delegates: <LocalizationsDelegate<dynamic>>[
+        DefaultWidgetsLocalizations.delegate,
+        DefaultMaterialLocalizations.delegate,
+      ],
+      child: new MediaQuery(
+        data: new MediaQueryData.fromWindow(window),
+        child: new Directionality(
```
Comment (Hixie) [inline packages/flutter/test/material/dropdown_test.dart]: ditto

File: packages/flutter_localizations/lib/src/l10n/material_en.arb [hunk 9]
```diff
@@ -154,4 +154,9 @@
   "@timePickerMinuteModeAnnouncement": {
     "description": "The audio announcement made when the time picker dialog is set to minute mode."
   }
+
+  "modalBarrierDismissLabel": "Dismiss",
+  "@modalBarrierDismissLabel": {
+    "description": "Accessibility label used to annotate a dismissible modal barrier."
```
Comment (Hixie) [inline packages/flutter_localizations/lib/src/l10n/material_en.arb]: give an example, e.g. "as seen behind a popup dialog box" or something

Comment (HansMuller) [inline packages/flutter_localizations/lib/src/l10n/material_en.arb]: This description will probably not help translators. You could include some of the text you wrote for routes.dart.

**Review by HansMuller — COMMENTED** [review]
You need to update flutter_localizations/lib/src/l10n/localizaions.dart, see https://github.com/flutter/flutter/blob/master/dev/tools/gen_localizations.dart

You just have to run:
```
dart dev/tools/gen_localizations.dart --overwrite
```

File: packages/flutter_localizations/lib/src/l10n/material_de_CH.arb (hunk 4 above)
Comment (goderbauer) [reply packages/flutter_localizations/lib/src/l10n/material_de_CH.arb]: Ack.

**Comment by goderbauer** [comment]
Addressed all comments. PTAL.

File: packages/flutter/test/material/dropdown_test.dart (hunk 8 above)
Comment (goderbauer) [reply packages/flutter/test/material/dropdown_test.dart]: This one is awkward to remove. The test verifies that things work with different textDirections. While Localizations does provide a text direction, I cannot modify it directly. I could rewrite the test to work on different locales instead, but I think that would hide the real purpose of the test (which is to make sure that different text directions work). Therefore, I left the Directionality widget in the tree.

Comment (Hixie) [reply packages/flutter/test/material/dropdown_test.dart]: that's fine. i didn't notice it was being tweaked.

File: packages/flutter/lib/src/widgets/modal_barrier.dart (hunk 3 above)
Comment (Hixie) [reply packages/flutter/lib/src/widgets/modal_barrier.dart]: ping

File: packages/flutter/lib/src/widgets/modal_barrier.dart (hunk 6 above)
Comment (Hixie) [reply packages/flutter/lib/src/widgets/modal_barrier.dart]: ping

**Comment by Hixie** [comment]
LGTM

**Pull request merged**